File: src/open_workshop_storage/utils/archive.py
from __future__ import annotations

import errno
import os
import pty
import re
import select
import shutil
import subprocess
import sys
from typing import Any, Callable, Optional


SEVEN_ZIP_BIN = "7z"
SEVEN_ZIP_PROGRESS_RE = re.compile(r"(?<!\d)(100|[1-9]?\d)%")
ZIP_MIN_COMPRESSION_SAVINGS_RATIO = 0.01


def ensure_7z_available() -> None:
    if shutil.which(SEVEN_ZIP_BIN) is None:
        raise RuntimeError("7z binary is required but not found in PATH")


def _resolve_tools_export(name: str, fallback: Callable[..., Any]) -> Callable[..., Any]:
    tools_module = sys.modules.get("open_workshop_storage.utils")
    if tools_module is None:
        return fallback
    override = getattr(tools_module, name, fallback)
    if callable(override):
        return override
    return fallback


def _run_7z(args: list[str], cwd: Optional[str] = None) -> subprocess.CompletedProcess:
    _resolve_tools_export("ensure_7z_available", ensure_7z_available)()
    return subprocess.run(
        [SEVEN_ZIP_BIN, *args],
        cwd=cwd,
        capture_output=True,
        text=True,
    )


def _drain_7z_progress_stream(
    chunk: str,
    state: dict[str, Any],
    on_progress: Optional[Callable[[int], None]] = None,
) -> None:
    pending = str(state.get("pending", "")) + chunk
    if len(pending) > 256:
        pending = pending[-256:]
    last_percent = int(state.get("last_percent", -1))
    for match in SEVEN_ZIP_PROGRESS_RE.finditer(pending):
        percent = int(match.group(1))
        if percent <= last_percent:
            continue
        last_percent = percent
        if on_progress is not None:
            on_progress(percent)
    state["pending"] = pending
    state["last_percent"] = last_percent


def _run_7z_with_progress(
    args: list[str],
    cwd: Optional[str] = None,
    on_progress: Optional[Callable[[int], None]] = None,
) -> subprocess.CompletedProcess:
    _resolve_tools_export("ensure_7z_available", ensure_7z_available)()
    master_fd, slave_fd = pty.openpty()
    process = subprocess.Popen(
        [SEVEN_ZIP_BIN, *args],
        cwd=cwd,
        stdin=subprocess.DEVNULL,
        stdout=slave_fd,
        stderr=slave_fd,
        close_fds=True,
    )
    os.close(slave_fd)

    output_chunks: list[str] = []
    progress_state = {"pending": "", "last_percent": -1}
    try:
        while True:
            ready, _, _ = select.select([master_fd], [], [], 0.1)
            if master_fd in ready:
                try:
                    data = os.read(master_fd, 4096)
                except OSError as exc:
                    if exc.errno == errno.EIO:
                        break
                    raise
                if not data:
                    break
                chunk = data.decode("utf-8", errors="replace")
                output_chunks.append(chunk)
                _drain_7z_progress_stream(chunk, progress_state, on_progress)
            if process.poll() is not None and master_fd not in ready:
                try:
                    data = os.read(master_fd, 4096)
                except OSError as exc:
                    if exc.errno == errno.EIO:
                        break
                    raise
                if not data:
                    break
                chunk = data.decode("utf-8", errors="replace")
                output_chunks.append(chunk)
                _drain_7z_progress_stream(chunk, progress_state, on_progress)
    finally:
        os.close(master_fd)

    returncode = process.wait()
    output = "".join(output_chunks)

    return subprocess.CompletedProcess(
        process.args,
        returncode,
        stdout=output,
        stderr="",
    )


def _run_7z_list(
    path: str,
    archive_type: Optional[str] = None,
) -> tuple[Optional[list[dict[str, str]]], str, int]:
    args = ["l", "-slt"]
    if archive_type:
        args.append(f"-t{archive_type}")
    args.append(path)
    result = _run_7z(args)
    if result.returncode != 0:
        output = (result.stderr or "") + "\n" + (result.stdout or "")
        return None, output.strip(), result.returncode
    entries: list[dict[str, str]] = []
    current: dict[str, str] = {}
    for line in result.stdout.splitlines():
        line = line.strip()
        if not line:
            if current:
                entries.append(current)
                current = {}
            continue
        if " = " in line:
            key, value = line.split(" = ", 1)
            current[key] = value
    if current:
        entries.append(current)
    return entries, "", 0


def _list_7z_entries(path: str, archive_type: Optional[str] = None) -> Optional[list[dict[str, str]]]:
    entries, _, code = _run_7z_list(path, archive_type=archive_type)
    if code != 0:
        return None
    return entries


def probe_archive(path: str) -> tuple[Optional[str], bool, Optional[list[dict[str, str]]]]:
    entries, error, code = _run_7z_list(path)
    if code != 0 or not entries:
        lowered = error.lower()
        if "password" in lowered or "encrypted" in lowered:
            return None, True, None
        return None, False, None
    archive_type = entries[0].get("Type")
    archive_type = archive_type.lower() if archive_type else None
    encrypted = False
    for entry in entries:
        if entry.get("Encrypted") == "+":
            encrypted = True
            break
    return archive_type, encrypted, entries


def archive_entries_unpacked_bytes(entries: Optional[list[dict[str, str]]]) -> Optional[int]:
    if entries is None:
        return None

    total = 0
    for entry in entries:
        if entry.get("Type"):
            continue
        if not entry.get("Path"):
            continue
        if entry.get("Folder") == "+":
            continue
        raw_size = entry.get("Size")
        if raw_size is None:
            return None
        try:
            size = int(raw_size)
        except (TypeError, ValueError):
            return None
        if size < 0:
            return None
        total += size
    return total


def archive_entries_packed_bytes(entries: Optional[list[dict[str, str]]]) -> Optional[int]:
    if entries is None:
        return None

    total = 0
    for entry in entries:
        if entry.get("Type"):
            continue
        if not entry.get("Path"):
            continue
        if entry.get("Folder") == "+":
            continue
        raw_size = entry.get("Packed Size")
        if raw_size is None:
            return None
        try:
            size = int(raw_size)
        except (TypeError, ValueError):
            return None
        if size < 0:
            return None
        total += size
    return total


def zip_dir_with_level(
    src_dir: str,
    dest_zip_path: str,
    compresslevel: int = 3,
    on_progress: Optional[Callable[[int], None]] = None,
) -> None:
    src_dir = os.path.abspath(src_dir)
    os.makedirs(os.path.dirname(dest_zip_path), exist_ok=True)
    if os.path.exists(dest_zip_path):
        os.remove(dest_zip_path)
    result = _run_7z_with_progress(
        [
            "a",
            "-tzip",
            "-mm=Deflate",
            f"-mx={compresslevel}",
            "-mmt=on",
            "-bb0",
            "-bso0",
            "-bsp1",
            dest_zip_path,
            ".",
        ],
        cwd=src_dir,
        on_progress=on_progress,
    )
    if result.returncode != 0:
        raise RuntimeError((result.stderr or result.stdout).strip() or "7z failed to create zip")


def zip_uses_deflated_or_better(
    path: str,
    entries: Optional[list[dict[str, str]]] = None,
) -> bool:
    entries = entries or _list_7z_entries(path, archive_type="zip")
    if not entries:
        return False

    for entry in entries:
        if entry.get("Type"):
            continue
        if not entry.get("Path"):
            continue
        if entry.get("Folder") == "+":
            continue
        if entry.get("Encrypted") == "+":
            return False
        method = (entry.get("Method") or "").lower()
        if not method:
            return False
        if "deflate" in method:
            continue
        if "lzma" in method or "bzip2" in method or "ppmd" in method:
            continue
        if "store" in method:
            continue
        return False

    unpacked_bytes = _resolve_tools_export(
        "archive_entries_unpacked_bytes",
        archive_entries_unpacked_bytes,
    )(entries)
    packed_bytes = _resolve_tools_export(
        "archive_entries_packed_bytes",
        archive_entries_packed_bytes,
    )(entries)
    if unpacked_bytes is None or packed_bytes is None:
        return False
    if unpacked_bytes <= 0:
        return True
    if packed_bytes >= unpacked_bytes:
        return False

    savings_ratio = (unpacked_bytes - packed_bytes) / unpacked_bytes
    return savings_ratio >= ZIP_MIN_COMPRESSION_SAVINGS_RATIO


def _find_single_tar(dest_dir: str) -> Optional[str]:
    entries = os.listdir(dest_dir)
    if len(entries) != 1:
        return None
    path = os.path.join(dest_dir, entries[0])
    if os.path.isfile(path) and entries[0].lower().endswith(".tar"):
        return path
    return None


def safe_extract_archive(
    archive_path: str,
    dest_dir: str,
    entries: Optional[list[dict[str, str]]] = None,
    on_progress: Optional[Callable[[int], None]] = None,
) -> None:
    dest_dir = os.path.abspath(dest_dir)
    entries = entries or _list_7z_entries(archive_path)
    if entries is None:
        raise ValueError("Invalid archive")
    for entry in entries:
        if entry.get("Encrypted") == "+":
            raise ValueError("Encrypted archive entries are not supported")
        if entry.get("Type"):
            continue
        name = (entry.get("Path") or "").replace("\\", "/")
        if not name:
            continue
        target_path = os.path.abspath(os.path.join(dest_dir, name))
        if os.path.commonpath([target_path, dest_dir]) != dest_dir:
            raise ValueError("Unsafe path in archive")
    os.makedirs(dest_dir, exist_ok=True)
    result = _run_7z_with_progress(
        [
            "x",
            f"-o{dest_dir}",
            "-y",
            "-bb0",
            "-bso0",
            "-bsp1",
            archive_path,
        ],
        on_progress=on_progress,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "7z failed to extract archive")

    archive_type = (entries[0].get("Type") or "").lower()
    if archive_type in {"gzip", "bzip2", "xz"}:
        tar_path = _find_single_tar(dest_dir)
        if tar_path:
            _resolve_tools_export("safe_extract_archive", safe_extract_archive)(
                tar_path,
                dest_dir,
                on_progress=on_progress,
            )
            os.remove(tar_path)

File: src/archive.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::time::Duration;

use regex::Regex;
use tokio::io::AsyncReadExt;
use tokio::process::Command;
use tokio::time::{sleep_until, Instant};

use crate::fs_utils::safe_path;

pub const SEVEN_ZIP_BIN: &str = "7z";
pub const ZIP_MIN_COMPRESSION_SAVINGS_RATIO: f64 = 0.01;
pub const DEFAULT_SEVEN_ZIP_TIMEOUT_SECONDS: u64 = 3600;
pub const DEFAULT_SEVEN_ZIP_IDLE_TIMEOUT_SECONDS: u64 = 60;
pub type ArchiveEntry = HashMap<String, String>;

#[derive(Debug, thiserror::Error)]
pub enum SevenZipError {
    #[error("7z binary is required but not found in PATH")]
    MissingBinary,
    #[error("7z timed out{suffix}")]
    Timeout { suffix: String },
    #[error("7z idle timed out{suffix}")]
    IdleTimeout { suffix: String },
    #[error("{0}")]
    CommandFailed(String),
    #[error("{0}")]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone)]
pub struct SevenZipOutput {
    pub status: std::process::ExitStatus,
    pub stdout: String,
    pub stderr: String,
}

pub fn ensure_7z_available() -> Result<(), SevenZipError> {
    let status = std::process::Command::new(SEVEN_ZIP_BIN)
        .arg("--help")
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status();
    match status {
        Ok(_) => Ok(()),
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Err(SevenZipError::MissingBinary),
        Err(err) => Err(SevenZipError::Io(err)),
    }
}

async fn run_7z_capture(
    args: &[String],
    cwd: Option<&Path>,
    timeout: Option<Duration>,
) -> Result<SevenZipOutput, SevenZipError> {
    ensure_7z_available()?;
    let mut command = Command::new(SEVEN_ZIP_BIN);
    command.args(args);
    command.stdin(Stdio::null());
    command.stdout(Stdio::piped());
    command.stderr(Stdio::piped());
    if let Some(cwd) = cwd {
        command.current_dir(cwd);
    }
    let output = match timeout {
        Some(timeout) => tokio::time::timeout(timeout, command.output())
            .await
            .map_err(|_| SevenZipError::Timeout {
                suffix: timeout_suffix(timeout),
            })??,
        None => command.output().await?,
    };
    Ok(SevenZipOutput {
        status: output.status,
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
    })
}

fn drain_progress_stream(
    chunk: &str,
    pending: &mut String,
    last_percent: &mut i32,
    on_progress: Option<&(dyn Fn(u32) + Send + Sync)>,
) {
    pending.push_str(chunk);
    if pending.len() > 256 {
        let start = pending.len() - 256;
        *pending = pending[start..].to_string();
    }
    static PROGRESS_RE: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
        Regex::new(r"(?:^|[^\d])(100|[1-9]?\d)%").expect("valid regex")
    });
    for caps in PROGRESS_RE.captures_iter(pending) {
        if let Ok(percent) = caps
            .get(1)
            .map(|m| m.as_str())
            .unwrap_or_default()
            .parse::<i32>()
        {
            if percent > *last_percent {
                *last_percent = percent;
                if let Some(cb) = on_progress {
                    cb(percent as u32);
                }
            }
        }
    }
}

async fn run_7z_with_progress(
    args: &[String],
    cwd: Option<&Path>,
    on_progress: Option<&(dyn Fn(u32) + Send + Sync)>,
    timeout: Option<Duration>,
    idle_timeout: Option<Duration>,
) -> Result<SevenZipOutput, SevenZipError> {
    ensure_7z_available()?;
    let mut command = Command::new(SEVEN_ZIP_BIN);
    command.args(args);
    command.stdin(Stdio::null());
    command.stdout(Stdio::piped());
    command.stderr(Stdio::piped());
    if let Some(cwd) = cwd {
        command.current_dir(cwd);
    }

    let mut child = command.spawn()?;
    let mut stdout = child
        .stdout
        .take()
        .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::Other, "missing stdout"))?;
    let stderr = child
        .stderr
        .take()
        .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::Other, "missing stderr"))?;

    let stderr_handle = tokio::spawn(async move {
        let mut stderr = stderr;
        let mut buf = Vec::new();
        let _ = stderr.read_to_end(&mut buf).await;
        buf
    });

    let mut stdout_bytes = Vec::new();
    let mut pending = String::new();
    let mut last_percent = -1;
    let total_deadline = timeout.map(|dur| Instant::now() + dur);
    let total_timeout_suffix = timeout.map(timeout_suffix);
    let idle_timeout_suffix = idle_timeout.map(timeout_suffix);
    let mut idle_deadline = idle_timeout.map(|dur| Instant::now() + dur);
    let mut buf = [0u8; 4096];

    loop {
        let deadline = match (total_deadline, idle_deadline) {
            (Some(a), Some(b)) => Some(a.min(b)),
            (Some(a), None) => Some(a),
            (None, Some(b)) => Some(b),
            (None, None) => None,
        };
        if let Some(deadline) = deadline {
            tokio::select! {
                read = stdout.read(&mut buf) => {
                    let n = read?;
                    if n == 0 {
                        break;
                    }
                    let chunk = String::from_utf8_lossy(&buf[..n]).to_string();
                    stdout_bytes.extend_from_slice(&buf[..n]);
                    drain_progress_stream(&chunk, &mut pending, &mut last_percent, on_progress);
                    if let Some(idle_timeout) = idle_timeout {
                        idle_deadline = Some(Instant::now() + idle_timeout);
                    }
                }
                _ = sleep_until(deadline) => {
                    let _ = child.kill().await;
                    return Err(match (total_deadline, idle_deadline) {
                        (Some(total_deadline), Some(idle_deadline)) if total_deadline <= idle_deadline => {
                            SevenZipError::Timeout {
                                suffix: total_timeout_suffix.clone().unwrap_or_default(),
                            }
                        }
                        (Some(_), Some(_)) => SevenZipError::IdleTimeout {
                            suffix: idle_timeout_suffix.clone().unwrap_or_default(),
                        },
                        (Some(_), None) => SevenZipError::Timeout {
                            suffix: total_timeout_suffix.clone().unwrap_or_default(),
                        },
                        (None, Some(_)) => SevenZipError::IdleTimeout {
                            suffix: idle_timeout_suffix.clone().unwrap_or_default(),
                        },
                        (None, None) => unreachable!("deadline fired without timeout"),
                    });
                }
            }
        } else {
            let n = stdout.read(&mut buf).await?;
            if n == 0 {
                break;
            }
            let chunk = String::from_utf8_lossy(&buf[..n]).to_string();
            stdout_bytes.extend_from_slice(&buf[..n]);
            drain_progress_stream(&chunk, &mut pending, &mut last_percent, on_progress);
            if let Some(idle_timeout) = idle_timeout {
                idle_deadline = Some(Instant::now() + idle_timeout);
            }
        }
    }

    let status = child.wait().await?;
    let stderr_bytes = stderr_handle.await.unwrap_or_default();
    Ok(SevenZipOutput {
        status,
        stdout: String::from_utf8_lossy(&stdout_bytes).to_string(),
        stderr: String::from_utf8_lossy(&stderr_bytes).to_string(),
    })
}

fn parse_list_output(stdout: &str) -> Vec<ArchiveEntry> {
    let mut entries = Vec::new();
    let mut current = ArchiveEntry::new();
    for line in stdout.lines() {
        let line = line.trim();
        if line.is_empty() {
            if !current.is_empty() {
                entries.push(current);
                current = ArchiveEntry::new();
            }
            continue;
        }
        if let Some((key, value)) = line.split_once(" = ") {
            current.insert(key.to_string(), value.to_string());
        }
    }
    if !current.is_empty() {
        entries.push(current);
    }
    entries
}

pub async fn run_7z_list(
    path: impl AsRef<Path>,
    archive_type: Option<&str>,
    timeout: Option<Duration>,
) -> Result<(Vec<ArchiveEntry>, String, i32), SevenZipError> {
    let mut args = vec!["l".to_string(), "-slt".to_string()];
    if let Some(archive_type) = archive_type {
        args.push(format!("-t{archive_type}"));
    }
    args.push(path.as_ref().to_string_lossy().to_string());
    let output = run_7z_capture(&args, None, timeout).await?;
    let code = output.status.code().unwrap_or(-1);
    if code != 0 {
        let error = format!("{}\n{}", output.stderr, output.stdout)
            .trim()
            .to_string();
        return Ok((Vec::new(), error, code));
    }
    Ok((parse_list_output(&output.stdout), String::new(), code))
}

pub async fn probe_archive(
    path: impl AsRef<Path>,
    timeout: Option<Duration>,
) -> Result<(Option<String>, bool, Option<Vec<ArchiveEntry>>), SevenZipError> {
    let (entries, error, code) = run_7z_list(path, None, timeout).await?;
    if code != 0 || entries.is_empty() {
        let lowered = error.to_lowercase();
        if lowered.contains("password") || lowered.contains("encrypted") {
            return Ok((None, true, None));
        }
        return Ok((None, false, None));
    }
    let archive_type = entries[0].get("Type").map(|value| value.to_lowercase());
    let encrypted = entries.iter().any(|entry| {
        entry
            .get("Encrypted")
            .map(|value| value == "+")
            .unwrap_or(false)
    });
    Ok((archive_type, encrypted, Some(entries)))
}

pub fn archive_entries_unpacked_bytes(entries: Option<&[ArchiveEntry]>) -> Option<u64> {
    let entries = entries?;
    let mut total = 0u64;
    for entry in entries {
        if entry.contains_key("Type") {
            continue;
        }
        if !entry.contains_key("Path") {
            continue;
        }
        if entry
            .get("Folder")
            .map(|value| value == "+")
            .unwrap_or(false)
        {
            continue;
        }
        let size = entry.get("Size")?.parse::<i64>().ok()?;
        if size < 0 {
            return None;
        }
        total += size as u64;
    }
    Some(total)
}

pub fn archive_entries_packed_bytes(entries: Option<&[ArchiveEntry]>) -> Option<u64> {
    let entries = entries?;
    let mut total = 0u64;
    for entry in entries {
        if entry.contains_key("Type") {
            continue;
        }
        if !entry.contains_key("Path") {
            continue;
        }
        if entry
            .get("Folder")
            .map(|value| value == "+")
            .unwrap_or(false)
        {
            continue;
        }
        let size = entry.get("Packed Size")?.parse::<i64>().ok()?;
        if size < 0 {
            return None;
        }
        total += size as u64;
    }
    Some(total)
}

pub fn zip_uses_deflated_or_better(entries: Option<&[ArchiveEntry]>) -> bool {
    let Some(entries) = entries else {
        return false;
    };
    for entry in entries {
        if entry.contains_key("Type") {
            continue;
        }
        if !entry.contains_key("Path") {
            continue;
        }
        if entry
            .get("Folder")
            .map(|value| value == "+")
            .unwrap_or(false)
        {
            continue;
        }
        if entry
            .get("Encrypted")
            .map(|value| value == "+")
            .unwrap_or(false)
        {
            return false;
        }
        let method = entry
            .get("Method")
            .map(|value| value.to_lowercase())
            .unwrap_or_default();
        if method.is_empty() {
            return false;
        }
        if method.contains("deflate")
            || method.contains("lzma")
            || method.contains("bzip2")
            || method.contains("ppmd")
            || method.contains("store")
        {
            continue;
        }
        return false;
    }

    let unpacked_bytes = archive_entries_unpacked_bytes(Some(entries));
    let packed_bytes = archive_entries_packed_bytes(Some(entries));
    let (Some(unpacked_bytes), Some(packed_bytes)) = (unpacked_bytes, packed_bytes) else {
        return false;
    };
    if unpacked_bytes == 0 {
        return true;
    }
    if packed_bytes >= unpacked_bytes {
        return false;
    }
    let savings_ratio = (unpacked_bytes - packed_bytes) as f64 / unpacked_bytes as f64;
    savings_ratio >= ZIP_MIN_COMPRESSION_SAVINGS_RATIO
}

fn find_single_tar(dest_dir: impl AsRef<Path>) -> Option<PathBuf> {
    let mut entries = std::fs::read_dir(dest_dir)
        .ok()?
        .filter_map(|entry| entry.ok())
        .collect::<Vec<_>>();
    if entries.len() != 1 {
        return None;
    }
    let entry = entries.pop()?;
    let path = entry.path();
    if path.is_file()
        && path
            .extension()
            .and_then(|ext| ext.to_str())
            .map(|ext| ext.eq_ignore_ascii_case("tar"))
            .unwrap_or(false)
    {
        Some(path)
    } else {
        None
    }
}

pub async fn zip_dir_with_level(
    src_dir: impl AsRef<Path>,
    dest_zip_path: impl AsRef<Path>,
    compresslevel: u32,
    on_progress: Option<&(dyn Fn(u32) + Send + Sync)>,
    timeout: Option<Duration>,
    idle_timeout: Option<Duration>,
) -> Result<(), SevenZipError> {
    let src_dir = src_dir.as_ref().to_path_buf();
    let dest_zip_path = dest_zip_path.as_ref().to_path_buf();
    if let Some(parent) = dest_zip_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    if dest_zip_path.exists() {
        std::fs::remove_file(&dest_zip_path)?;
    }
    let args = vec![
        "a".to_string(),
        "-tzip".to_string(),
        "-mm=Deflate".to_string(),
        format!("-mx={compresslevel}"),
        "-mmt=on".to_string(),
        "-bb0".to_string(),
        "-bso0".to_string(),
        "-bsp1".to_string(),
        dest_zip_path.to_string_lossy().to_string(),
        ".".to_string(),
    ];
    let output =
        run_7z_with_progress(&args, Some(&src_dir), on_progress, timeout, idle_timeout).await?;
    if !output.status.success() {
        let error = if output.stderr.trim().is_empty() {
            output.stdout.trim()
        } else {
            output.stderr.trim()
        };
        return Err(SevenZipError::CommandFailed(error.to_string()));
    }
    Ok(())
}

pub async fn safe_extract_archive(
    archive_path: impl AsRef<Path>,
    dest_dir: impl AsRef<Path>,
    entries: Option<&[ArchiveEntry]>,
    on_progress: Option<&(dyn Fn(u32) + Send + Sync)>,
    timeout: Option<Duration>,
    idle_timeout: Option<Duration>,
) -> Result<(), SevenZipError> {
    let mut archive_path = archive_path.as_ref().to_path_buf();
    let dest_dir = dest_dir.as_ref().to_path_buf();
    let mut entries = entries.map(|entries| entries.to_vec());
    let mut remove_after_extract: Option<PathBuf> = None;

    loop {
        let current_entries = if let Some(entries) = entries.as_ref() {
            entries.as_slice()
        } else {
            let (listed_entries, error, code) = run_7z_list(&archive_path, None, timeout).await?;
            if code != 0 || listed_entries.is_empty() {
                let lowered = error.to_lowercase();
                if lowered.contains("password") || lowered.contains("encrypted") {
                    return Err(SevenZipError::CommandFailed(
                        "Encrypted archive entries are not supported".to_string(),
                    ));
                }
                return Err(SevenZipError::CommandFailed(if error.trim().is_empty() {
                    "Invalid archive".to_string()
                } else {
                    error
                }));
            }
            entries = Some(listed_entries);
            entries.as_ref().unwrap().as_slice()
        };

        for entry in current_entries {
            if entry
                .get("Encrypted")
                .map(|value| value == "+")
                .unwrap_or(false)
            {
                return Err(SevenZipError::CommandFailed(
                    "Encrypted archive entries are not supported".to_string(),
                ));
            }
            if entry.contains_key("Type") {
                continue;
            }
            let name = entry
                .get("Path")
                .cloned()
                .unwrap_or_default()
                .replace('\\', "/");
            if name.is_empty() {
                continue;
            }
            let _ = safe_path(&dest_dir, &name).map_err(SevenZipError::Io)?;
        }

        std::fs::create_dir_all(&dest_dir)?;
        let args = vec![
            "x".to_string(),
            format!("-o{}", dest_dir.to_string_lossy()),
            "-y".to_string(),
            "-bb0".to_string(),
            "-bso0".to_string(),
            "-bsp1".to_string(),
            archive_path.to_string_lossy().to_string(),
        ];
        let output = run_7z_with_progress(&args, None, on_progress, timeout, idle_timeout).await?;
        if !output.status.success() {
            let error = if output.stderr.trim().is_empty() {
                output.stdout.trim()
            } else {
                output.stderr.trim()
            };
            return Err(SevenZipError::CommandFailed(error.to_string()));
        }

        if let Some(path) = remove_after_extract.take() {
            std::fs::remove_file(path)?;
        }

        let archive_type = current_entries[0]
            .get("Type")
            .map(|value| value.to_lowercase())
            .unwrap_or_default();
        if ["gzip", "bzip2", "xz"].contains(&archive_type.as_str()) {
            if let Some(tar_path) = find_single_tar(&dest_dir) {
                remove_after_extract = Some(tar_path.clone());
                archive_path = tar_path;
                entries = None;
                continue;
            }
        }
        break;
    }
    Ok(())
}

fn timeout_suffix(timeout: Duration) -> String {
    format!(" after {}s", timeout.as_secs_f64())
}
